contracts/contracts/FleekPausable.sol

```diff
@@ -0,0 +1,134 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.7;
+
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+
+error ContractIsPaused();
+error ContractIsNotPaused();
+error ContractIsNotPausable();
+error PausableIsSetTo(bool state);
+
+abstract contract FleekPausable is Initializable {
+    /**
+     * @dev Emitted when the pause is triggered by `account` and set to `isPaused`.
+     */
+    event PauseStatusChange(bool indexed isPaused, address account);
+
+    /**
+     * @dev Emitted when the pausable is triggered by `account` and set to `isPausable`.
+     */
+    event PausableStatusChange(bool indexed isPausable, address account);
+
+    bool private _paused;
+    bool private _canPause; // TODO: how should we verify if the contract is pausable or not?
+
+    /**
+     * @dev Initializes the contract in paused state.
+     */
+    function __FleekPausable_init() internal onlyInitializing {
+        _paused = true;
+        _canPause = true;
+    }
+
+    /**
+     * @dev Modifier to make a function callable only when the contract is not paused.
+     *
+     * Requirements:
+     *
+     * - The contract must not be paused.
+     */
+    modifier whenNotPaused() {
+        _requireNotPaused();
+        _;
+    }
+
+    /**
+     * @dev Modifier to make a function callable only when the contract is paused.
+     *
+     * Requirements:
+     *
+     * - The contract must be paused.
+     */
+    modifier whenPaused() {
+        _requirePaused();
+        _;
+    }
+
+    /**
+     * @dev Returns true if the contract is paused, and false otherwise.
+     */
+    function isPaused() public view returns (bool) {
+        return _paused;
+    }
+
+    /**
+     * @dev Returns true if the contract is pausable, and false otherwise.
+     */
+    function isPausable() public view returns (bool) {
+        return _canPause;
+    }
+
+    /**
+     * @dev Throws if the contract is paused.
+     */
+    function _requireNotPaused() internal view {
+        if (isPaused()) revert ContractIsPaused();
+    }
+
+    /**
+     * @dev Throws if the contract is not paused.
+     */
+    function _requirePaused() internal view {
+        if (!isPaused()) revert ContractIsNotPaused();
+    }
+
+    /**
+     * @dev Throws if the contract is not pausable.
+     */
+    function _requirePausable() internal view {
+        if (!isPausable()) revert ContractIsNotPausable();
+    }
+
+    /**
+     * @dev Sets the contract to be pausable or not.
+     * @param canPause true if the contract is pausable, and false otherwise.
+     */
+    function _setPausable(bool canPause) internal {
```

Comment:
Should we do this like a toggle where we can switch it back too? Or create a separate function that sets canPause to false. This way its a feature flag that could be set back and forth as governance decides (and at the moment Fleek controller).

Reply:
I think I did not get the question. The function should always invert the state? I just think this way is safer because you cannot set the flag to a state you don't want by accident.

Reply:
But we should still have a way to turn of off after. Right now once you make it pausable you can never go back?

Reply:
No, we can change it whenever we want. It just checks if you are trying to set the same value as it is, in this case it reverts.